File: backend/src/database/seed-geography.ts
/**
 * Seed géographique — port TypeScript de database/seed_geography.py
 * (voir ce fichier pour le contexte complet et les sources).
 *
 * Ce port existe uniquement pour que le déploiement (Railway ou autre)
 * n'ait pas besoin de python3/psycopg2 dans l'image : le backend est déjà
 * en Node, donc réutiliser `pg` (déjà une dépendance) évite un second
 * runtime au moment du build. La logique — y compris les deux correctifs
 * de qualité de données — est identique à la version Python :
 *   1. Les champs "longitude"/"latitude" de la source sont inversés.
 *   2. Deux communes ont un champ numérique malformé (virgule parasite).
 */
import { Client } from 'pg';
import * as fs from 'fs';
import * as path from 'path';

interface RawWilaya {
  id: string;
  code: string;
  name: string;
  ar_name: string;
  longitude: string; // contient en réalité la LATITUDE (bug de la source)
  latitude: string; // contient en réalité la LONGITUDE (bug de la source)
}

interface RawCommune {
  id: string;
  post_code: string;
  name: string;
  wilaya_id: string;
  ar_name: string;
  longitude: string;
  latitude: string;
}

function readJsonStrippingBom(filePath: string) {
  const raw = fs.readFileSync(filePath, 'utf-8');
  const stripped = raw.charCodeAt(0) === 0xfeff ? raw.slice(1) : raw;
  return JSON.parse(stripped);
}

/** Nettoie les valeurs numériques mal formées (ex. ",2.8140852") plutôt que de planter. */
function safeFloat(value: string | undefined | null): number | null {
  if (value === undefined || value === null) return null;
  const cleaned = String(value).trim().replace(/^,+/, '');
  const parsed = parseFloat(cleaned);
  return Number.isNaN(parsed) ? null : parsed;
}

export async function seedGeography(client: Client) {
  const dbDir = path.join(process.cwd(), 'database');
  const wilayas: RawWilaya[] = readJsonStrippingBom(path.join(dbDir, 'wilayas_raw.json'));
  const communes: RawCommune[] = readJsonStrippingBom(path.join(dbDir, 'communes_raw.json'));

  console.log(`[seed] ${wilayas.length} wilayas, ${communes.length} lignes de communes brutes`);

  await client.query('TRUNCATE communes, dairas, wilayas RESTART IDENTITY CASCADE');

  const wilayaIdMap = new Map<string, number>();
  for (const w of wilayas) {
    const code = String(w.code).padStart(2, '0');
    const realLat = safeFloat(w.longitude); // inversion volontaire, voir en-tête
    const realLng = safeFloat(w.latitude);
    const res = await client.query(
      `INSERT INTO wilayas (code, name_fr, name_ar, center)
       VALUES ($1, $2, $3, ST_SetSRID(ST_MakePoint($4, $5), 4326)::geography)
       RETURNING id`,
      [code, w.name.trim(), w.ar_name.trim(), realLng, realLat],
    );
    wilayaIdMap.set(String(w.id), res.rows[0].id);
  }
  console.log(`[seed] ${wilayaIdMap.size} wilayas insérées`);

  const seen = new Map<string, RawCommune>();
  for (const c of communes) {
    const key = `${c.name.trim().toLowerCase()}|${c.wilaya_id}`;
    if (!seen.has(key)) seen.set(key, c);
  }
  console.log(`[seed] ${communes.length} -> ${seen.size} communes uniques après déduplication`);

  let inserted = 0;
  let skipped = 0;
  for (const c of seen.values()) {
    const wilayaDbId = wilayaIdMap.get(String(c.wilaya_id));
    if (!wilayaDbId) {
      skipped++;
      continue;
    }
    const realLat = safeFloat(c.longitude);
    const realLng = safeFloat(c.latitude);
    if (realLat === null || realLng === null) {
      console.warn(`[seed] AVERTISSEMENT : coordonnées illisibles pour "${c.name}" (id=${c.id}), location NULL`);
    }
    await client.query(
      `INSERT INTO communes (wilaya_id, postal_code, name_fr, name_ar, location)
       VALUES ($1, $2, $3, $4, ST_SetSRID(ST_MakePoint($5, $6), 4326)::geography)`,
      [wilayaDbId, c.post_code ?? null, c.name.trim(), (c.ar_name ?? '').trim() || null, realLng, realLat],
    );
    inserted++;
  }
  console.log(`[seed] ${inserted} communes insérées, ${skipped} ignorées (wilaya inconnue)`);
}

File: backend/database/seed_geography.py
#!/usr/bin/env python3
"""
Seed script — géographie administrative de l'Algérie (wilayas + communes).

Source des données brutes : dataset ouvert GitHub
  Kenandarabeh/algeria-wilayas-communes-2026 (69 wilayas, loi 26-06 avril 2026)

Nettoyage appliqué (et pourquoi) :
  1. Le fichier source inverse les champs "longitude" et "latitude"
     (ex. Adrar est donné à longitude=27.97/latitude=-0.20, alors que les
     coordonnées réelles d'Adrar sont lat≈27.87°N, lng≈-0.29°E). On corrige
     en permutant les deux valeurs à l'import.
  2. Le fichier communes.json contient 1708 lignes alors que le chiffre
     officiel attendu est 1541 communes. Après déduplication stricte sur
     (nom, wilaya) on obtient 1609 lignes uniques : il reste un écart de
     68 communes par rapport au chiffre officiel que cette source ne permet
     pas d'expliquer avec certitude (granularité "agglomération" vs
     "commune" possible). CE SCRIPT CHARGE LES 1609 LIGNES DÉDUPLIQUÉES ET
     LE MARQUE EXPLICITEMENT — à ne pas considérer comme validé pour la
     production sans recoupement avec la nomenclature officielle ONS /
     Journal Officiel. Voir docs/DATA_MODEL.md.

Usage:
    python3 seed_geography.py
"""
import json
import os
import psycopg2

DB_DSN = os.environ.get("DATABASE_URL", "postgresql://postgres:devpassword@localhost:5432/covoit_dz")

HERE = os.path.dirname(os.path.abspath(__file__))


def safe_float(value):
    """Certaines lignes de la source contiennent des valeurs numériques
    mal formées (ex. ',2.8140852' avec une virgule parasite en tête).
    On nettoie plutôt que d'échouer silencieusement ou de planter."""
    if value is None:
        return None
    s = str(value).strip().lstrip(",")
    try:
        return float(s)
    except ValueError:
        return None


def load(path):
    with open(os.path.join(HERE, path), encoding="utf-8-sig") as f:
        return json.load(f)


def main():
    wilayas = load("../../wilayas_raw.json") if os.path.exists(os.path.join(HERE, "../../wilayas_raw.json")) else load("wilayas_raw.json")
    communes = load("../../communes_raw.json") if os.path.exists(os.path.join(HERE, "../../communes_raw.json")) else load("communes_raw.json")

    conn = psycopg2.connect(DB_DSN)
    conn.autocommit = False
    cur = conn.cursor()

    print(f"Loaded {len(wilayas)} wilayas, {len(communes)} raw commune rows")

    cur.execute("TRUNCATE communes, dairas, wilayas RESTART IDENTITY CASCADE;")

    wilaya_id_map = {}  # source_id (string) -> db id
    for w in wilayas:
        code = str(w["code"]).zfill(2)
        name_fr = w["name"].strip()
        name_ar = w["ar_name"].strip()
        # champs inversés dans la source : "longitude" contient la latitude réelle
        real_lat = float(w["longitude"])
        real_lng = float(w["latitude"])
        cur.execute(
            """
            INSERT INTO wilayas (code, name_fr, name_ar, center)
            VALUES (%s, %s, %s, ST_SetSRID(ST_MakePoint(%s, %s), 4326)::geography)
            RETURNING id
            """,
            (code, name_fr, name_ar, real_lng, real_lat),
        )
        db_id = cur.fetchone()[0]
        wilaya_id_map[str(w["id"])] = db_id

    print(f"Inserted {len(wilaya_id_map)} wilayas")

    # Dédoublonnage strict sur (nom normalisé, wilaya_id)
    seen = {}
    for c in communes:
        key = (c["name"].strip().lower(), c["wilaya_id"])
        if key not in seen:
            seen[key] = c
    unique_communes = list(seen.values())
    print(f"Deduplicated to {len(unique_communes)} unique communes")

    inserted, skipped = 0, 0
    for c in unique_communes:
        wilaya_db_id = wilaya_id_map.get(str(c["wilaya_id"]))
        if wilaya_db_id is None:
            skipped += 1
            continue
        name_fr = c["name"].strip()
        name_ar = (c.get("ar_name") or "").strip() or None
        postal_code = c.get("post_code")
        # même correction d'inversion longitude/latitude que pour les wilayas
        real_lat = safe_float(c["longitude"])
        real_lng = safe_float(c["latitude"])
        if real_lat is None or real_lng is None:
            print(f"  WARNING: skipping geo point for '{name_fr}' (id={c['id']}) — unparseable coordinates, inserting NULL location")
        cur.execute(
            """
            INSERT INTO communes (wilaya_id, postal_code, name_fr, name_ar, location)
            VALUES (%s, %s, %s, %s, ST_SetSRID(ST_MakePoint(%s, %s), 4326)::geography)
            """,
            (wilaya_db_id, postal_code, name_fr, name_ar, real_lng, real_lat),
        )
        inserted += 1

    print(f"Inserted {inserted} communes, skipped {skipped} (unknown wilaya)")

    conn.commit()
    cur.close()
    conn.close()
    print("Done.")


if __name__ == "__main__":
    main()
